crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -40,24 +46,160 @@ pub use bevy_ecs_macros::States;
 /// }
 ///
 /// ```
-pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {
-    /// The type returned when iterating over all [`variants`](States::variants) of this type.
-    type Iter: Iterator<Item = Self>;
+pub trait States: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug + Default {}
+
+/// Types that can match world-wide states.
+pub trait StateMatcher<S: States>: 'static + Send + Sync + Clone {
+    /// Check whether to match with the current state
+    fn match_state(&self, state: &S) -> bool;
+}
+
+impl<S: States> StateMatcher<S> for S {
+    fn match_state(&self, state: &S) -> bool {
```

Comment:
I don't immediately understand why we can't make this a required method on `States`, with a default impl.

Reply:
Oooh, that's a good idea, is that possible?

Reply:
If this were just part of the States object, then you would have to pre-determine all the ways you want to compare a state to other states in advance, and it would have to be present within that States object. That's fine if you're just relying on a limited set of mutually exclusive states (like the current enums provide with Eq). But it woudn't work with anything else.

So, for example, in a simple state like:
```rust
AppState {
Menu,
InGame { paused: bool }
}
```

You wouldn't be able to define something that matches all "InGame" states, regardless of whether the game is paused or not. Because you are always matching an incoming state to the current state, with no other information.

Now - you could still do "directional" matches, so "InGame { paused: true }" will accept "InGame { paused: false}" but not vice versa, but that doesn't cover most cases....

The StateMatcher trait is the thing that allows us to define a variety of ways to match against the same States object. There is already a default impl for States, which relies on Eq, but this is what allows us to:
- use closures (generated by the pattern matching macros)
- create marker objects that can serve as to match multiple states (so you could define a `InGame` struct, and implement the trait to make it match both paused and unpaused game states

Reply:
I think we could make it less accessible, for example make it so there are only 2 publicly implemented versions - the States built in one, and the StateMatcherFunction struct, which gets generated by passing in a function. However, that removes some of the re-usability that having the ability to define your own struct provides (see the advanced_states example for how I used it)